File: api/src/main/java/org/pmiops/workbench/audit/ActionAuditEventImpl.java
package org.pmiops.workbench.audit;

import java.util.Optional;

public class ActionAuditEventImpl implements ActionAuditEvent {

  private final long timestamp;
  private final AgentType agentType;
  private final long agentId;
  private final String actionId;
  private Optional<String> agentEmail;
  private final ActionType actionType;
  private final TargetType targetType;
  private final Optional<String> targetPropertyMaybe;
  private Optional<Long> targetIdMaybe;
  private final Optional<String> previousValueMaybe;
  private final Optional<String> newValueMaybe;

  public ActionAuditEventImpl(
      long timestamp,
      AgentType agentType,
      long agentId,
      Optional<String> agentEmail,
      String actionId,
      ActionType actionType,
      TargetType targetType,
      Optional<String> targetPropertyMaybe,
      Optional<Long> targetIdMaybe,
      Optional<String> previousValueMaybe,
      Optional<String> newValueMaybe) {
    this.timestamp = timestamp;
    this.agentType = agentType;
    this.agentId = agentId;
    this.agentEmail = agentEmail;
    this.actionId = actionId;
    this.actionType = actionType;
    this.targetType = targetType;
    this.targetPropertyMaybe = targetPropertyMaybe;
    this.targetIdMaybe = targetIdMaybe;
    this.previousValueMaybe = previousValueMaybe;
    this.newValueMaybe = newValueMaybe;
  }

  @Override
  public long timestamp() {
    return timestamp;
  }

  @Override
  public AgentType agentType() {
    return agentType;
  }

  @Override
  public long agentId() {
    return agentId;
  }

  @Override
  public Optional<String> agentEmail() {
    return agentEmail;
  }

  @Override
  public String actionId() {
    return actionId;
  }

  @Override
  public ActionType actionType() {
    return actionType;
  }

  @Override
  public TargetType targetType() {
    return targetType;
  }

  @Override
  public Optional<Long> targetId() {
    return targetIdMaybe; // check if this is null somehow
  }

  @Override
  public Optional<String> targetProperty() {
    return targetPropertyMaybe;
  }

  @Override
  public Optional<String> previousValue() {
    return previousValueMaybe;
  }

  @Override
  public Optional<String> newValue() {
    return newValueMaybe;
  }

  public static Builder builder() {
    return new Builder();
  }
  // This builder handles null inputs for optional fields, simplifying code at the call site.
  // For example setAgentEmail() can be called safely with null (or just not called it's not known).
  public static class Builder {

    private long timestamp;
    private AgentType agentType;
    private long agentId;
    private String agentEmail;
    private String actionId;
    private ActionType actionType;
    private TargetType targetType;
    private String targetProperty;
    private Long targetId;
    private String previousValue;
    private String newValue;

    public Builder() {}

    public Builder setTimestamp(long timestamp) {
      this.timestamp = timestamp;
      return this;
    }

    public Builder setAgentType(AgentType agentType) {
      this.agentType = agentType;
      return this;
    }

    public Builder setAgentId(long agentId) {
      this.agentId = agentId;
      return this;
    }

    public Builder setAgentEmail(String agentEmail) {
      this.agentEmail = agentEmail;
      return this;
    }

    public Builder setActionId(String actionId) {
      this.actionId = actionId;
      return this;
    }

    public Builder setActionType(ActionType actionType) {
      this.actionType = actionType;
      return this;
    }

    public Builder setTargetType(TargetType targetType) {
      this.targetType = targetType;
      return this;
    }

    public Builder setTargetProperty(String targetProperty) {
      this.targetProperty = targetProperty;
      return this;
    }

    public Builder setTargetId(Long targetId) {
      this.targetId = targetId;
      return this;
    }

    public Builder setPreviousValue(String previousValue) {
      this.previousValue = previousValue;
      return this;
    }

    public Builder setNewValue(String newValue) {
      this.newValue = newValue;
      return this;
    }

    public ActionAuditEventImpl build() {
      return new ActionAuditEventImpl(
          timestamp,
          agentType,
          agentId,
          Optional.ofNullable(agentEmail),
          actionId,
          actionType,
          targetType,
          Optional.ofNullable(targetProperty),
          Optional.ofNullable(targetId),
          Optional.ofNullable(previousValue),
          Optional.ofNullable(newValue));
    }
  }
}

File: api/src/main/java/org/pmiops/workbench/audit/ActionAuditEventImpl.kt
package org.pmiops.workbench.audit

import java.util.Optional
import org.elasticsearch.common.Strings

// This value class presents an Immutables-style builder
// and access pattern. All consturction must happen via the
// builder, and any missing required fields will generate an
// IllegalArgumentException. There are no public mutators on the fields.
class ActionAuditEventImpl private constructor(
        private val timestamp: Long,
        private val agentType: AgentType,
        private val agentId: Long,
        private val agentEmailMaybe: Optional<String>,
        private val actionId: String,
        private val actionType: ActionType,
        private val targetType: TargetType,
        private val targetPropertyMaybe: Optional<String>,
        private val targetIdMaybe: Optional<Long>,
        private val previousValueMaybe: Optional<String>,
        private val newValueMaybe: Optional<String>) : ActionAuditEvent {

    override fun timestamp(): Long {
        return timestamp
    }

    override fun agentType(): AgentType {
        return agentType
    }

    override fun agentId(): Long {
        return agentId
    }

    override fun agentEmail(): Optional<String> {
        return agentEmailMaybe
    }

    override fun actionId(): String {
        return actionId
    }

    override fun actionType(): ActionType {
        return actionType
    }

    override fun targetType(): TargetType {
        return targetType
    }

    override fun targetId(): Optional<Long> {
        return targetIdMaybe // check if this is null somehow
    }

    override fun targetProperty(): Optional<String> {
        return targetPropertyMaybe
    }

    override fun previousValue(): Optional<String> {
        return previousValueMaybe
    }

    override fun newValue(): Optional<String> {
        return newValueMaybe
    }

    // This builder handles null inputs for optional fields, simplifying code at the call site.
    // For example setAgentEmail() can be called safely with null (or just not called it's not known).
    class Builder {

        private var timestamp: Long = 0
        private var agentType: AgentType? = null
        private var agentId: Long = 0
        private var agentEmail: String? = null
        private var actionId: String? = null
        private var actionType: ActionType? = null
        private var targetType: TargetType? = null
        private var targetProperty: String? = null
        private var targetId: Long? = null
        private var previousValue: String? = null
        private var newValue: String? = null

        fun setTimestamp(timestamp: Long): Builder {
            this.timestamp = timestamp
            return this
        }

        fun setAgentType(agentType: AgentType): Builder {
            this.agentType = agentType
            return this
        }

        fun setAgentId(agentId: Long): Builder {
            this.agentId = agentId
            return this
        }

        fun setAgentEmail(agentEmail: String): Builder {
            this.agentEmail = agentEmail
            return this
        }

        fun setActionId(actionId: String): Builder {
            this.actionId = actionId
            return this
        }

        fun setActionType(actionType: ActionType): Builder {
            this.actionType = actionType
            return this
        }

        fun setTargetType(targetType: TargetType): Builder {
            this.targetType = targetType
            return this
        }

        fun setTargetProperty(targetProperty: String): Builder {
            this.targetProperty = targetProperty
            return this
        }

        fun setTargetId(targetId: Long?): Builder {
            this.targetId = targetId
            return this
        }

        fun setPreviousValue(previousValue: String): Builder {
            this.previousValue = previousValue
            return this
        }

        fun setNewValue(newValue: String): Builder {
            this.newValue = newValue
            return this
        }

        @Throws(IllegalArgumentException::class)
        private fun validateRequiredFields() {
            require(timestamp > 0L) { "Missing required field timestamp" }
            requireNotNull(agentType) { "Missing required field agentType" }
        }

        fun build(): ActionAuditEventImpl {
            validateRequiredFields()
            return ActionAuditEventImpl(
                    timestamp,
                    agentType,
                    agentId,
                    Optional.ofNullable(agentEmail),
                    actionId,
                    actionType,
                    targetType,
                    Optional.ofNullable(targetProperty),
                    Optional.ofNullable(targetId),
                    Optional.ofNullable(previousValue),
                    Optional.ofNullable(newValue))
        }
    }

    companion object {

        fun builder(): Builder {
            return Builder()
        }
    }
}
